core/src/main/java/io/grpc/transport/MessageDeframer.java

```diff
@@ -124,27 +145,136 @@ public MessageDeframer(Listener listener) {
   public MessageDeframer(Listener listener, Compression compression) {
     this.listener = Preconditions.checkNotNull(listener, "sink");
     this.compression = Preconditions.checkNotNull(compression, "compression");
+    queueRef.set(new ArrayDeque<Frame>(4));
   }
 
   /**
    * Requests up to the given number of messages from the call to be delivered to
-   * {@link Listener#messageRead(InputStream)}. No additional messages will be delivered.
+   * {@link Listener#messagesAvailable}. No additional messages will be delivered.
+   * Only called by the application thread.
    *
    * <p>If {@link #close()} has been called, this method will have no effect.
    *
    * @param numMessages the requested number of messages to be delivered to the listener.
    */
-  public void request(int numMessages) {
+  void request(int numMessages) {
     Preconditions.checkArgument(numMessages > 0, "numMessages must be > 0");
     if (isClosed()) {
       return;
     }
-    pendingDeliveries += numMessages;
-    deliver();
+    // Make the number of requested messages visible to the transport thread.
+    transportCount.addAndGet(numMessages);
+    requestedMessageCount += numMessages;
+    if (messagesAvailable && !appInRequest) {
+      appInRequest = true;
+      try {
+        listener.messagesAvailable(this);
+      } finally {
+        appInRequest = false;
+      }
+    }
   }
 
   /**
-   * Adds the given data to this deframer and attempts delivery to the sink.
+   * Called in the application thread to consume the available messages provided by the transport.
+   * @param consumer which receives all available messages.
+   */
+  @Override
+  public int drainTo(StreamListener.MessageConsumer consumer) {
+    int count = 0;
+    ArrayDeque<Frame> current;
+    do {
+      // Get the queue set by the transport, guaranteed to eventually be non-null
+      while ((current = queueRef.getAndSet(null)) == null) {
+        // Consider yielding if we are waiting too long for the transport to set a queue.
+      }
+      // Consume the queue
+      count += drainTo(consumer, current);
+      messagesAvailable = !current.isEmpty();
+
+      if (messagesAvailable) {
+        // We can't fully consume the queue yet so we have to accumulate anything produced
+        // by the transport thread while we were executing and return the merged version.
+        ArrayDeque<Frame> next = null;
+        while (!queueRef.compareAndSet(next, current)) {
+          next = queueRef.get();
```

Comment:
I think this should be a `getAndSet(null)`, otherwise the transport thread could be modifying the ArrayDeque while we are reading it here (among other issues).